# 16 - Final AquaPulse Results Summary

## Project Overview

AquaPulse combines:

- AquaSensor river measurements
- Environmental and weather information
- Dissolved oxygen forecasting
- Linear Regression, Random Forest and XGBoost
- Eight prediction horizons from 15 to 120 minutes
- Anomaly detection
- SHAP explainability
- Live monitoring
- Interactive Flask visualisation

This notebook provides a final summary of the main computational outputs produced during the project.

In [1]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd().parent

PERFORMANCE_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "do_prediction_model_performance.csv"
)

performance = pd.read_csv(
    PERFORMANCE_FILE
)

print("Total model evaluation records:", len(performance))

Total model evaluation records: 48


In [2]:
do_results = performance[
    performance["target_type"]
    .astype(str)
    .str.lower()
    == "do mg/l"
].copy()

best_results = (
    do_results
    .sort_values(
        ["horizon", "R2"],
        ascending=[True, False]
    )
    .groupby("horizon", as_index=False)
    .first()
)

best_results[
    [
        "horizon",
        "model",
        "MAE",
        "RMSE",
        "R2"
    ]
]

,horizon,model,MAE,RMSE,R2
0,105min,Linear Regression,0.195606,0.427643,0.947274
1,120min,Linear Regression,0.214726,0.456126,0.940018
2,15min,Linear Regression,0.059432,0.142753,0.994124
3,30min,Linear Regression,0.085146,0.200089,0.988456
4,45min,Linear Regression,0.108538,0.247535,0.982333
5,60min,Linear Regression,0.129906,0.282753,0.976948
6,75min,Linear Regression,0.152506,0.334699,0.967701
7,90min,Linear Regression,0.174986,0.388585,0.956464


In [3]:
print("Best model by forecast horizon:")

for _, row in best_results.iterrows():

    print(
        f"{row['horizon']}: "
        f"{row['model']} | "
        f"R²={row['R2']:.4f} | "
        f"RMSE={row['RMSE']:.4f}"
    )

Best model by forecast horizon:
105min: Linear Regression | R²=0.9473 | RMSE=0.4276
120min: Linear Regression | R²=0.9400 | RMSE=0.4561
15min: Linear Regression | R²=0.9941 | RMSE=0.1428
30min: Linear Regression | R²=0.9885 | RMSE=0.2001
45min: Linear Regression | R²=0.9823 | RMSE=0.2475
60min: Linear Regression | R²=0.9769 | RMSE=0.2828
75min: Linear Regression | R²=0.9677 | RMSE=0.3347
90min: Linear Regression | R²=0.9565 | RMSE=0.3886


## Main Project Achievements

The final AquaPulse system successfully:

1. Processed high-frequency AquaSensor data from River Derwent monitoring stations.
2. Integrated environmental and weather information.
3. Created DO forecasts for 15, 30, 45, 60, 75, 90, 105 and 120 minutes.
4. Compared Linear Regression, Random Forest and XGBoost.
5. Selected models using MAE, RMSE and R².
6. Implemented anomaly detection.
7. Applied SHAP explainability.
8. Integrated live AquaSensor information.
9. Developed an interactive Flask web application.
10. Presented actual DO, predicted DO and water temperature through interactive graphs.
11. Preserved historical live measurements for monitoring and analysis.